doc/mid-based-signalling.md

```diff
@@ -0,0 +1,276 @@
+that the first structure could be marshalled to JSON and sent to clients as-is.
+
+On the client side, the client will essentially take the `m.call.advertise` data and save it
+almost as-is. It would probably cross-reference it against the call member state events to
+ensure that it wasn't showing feeds for any users that did not have state events indicating
+that they were in the call, although if we trust the focus and assume that conf IDs are unique
+enough to be unguessable, this may be unnecessary.
+
+In the simplest case, the client will simply iterate through this structure and add every
+media UUID it finds to an `m.call.subscribe` message.
+
+The most complex part is that when the `m.call.describe` message arrives back from the focus,
+it will have to search through the data from the `m.call.advertise` message to map the tracks
+it is now receving to the right user IDs (XXX: it could build them into a map to make this lookup
+efficient, although if we make the advertise message indexed by media UUID then it already has
+a map indexed by the correct thing...)
+
+The `select_answer` is also tweaked to be more extensible-event like although is essentially
+the same:
+
+m.select\_answer
+```
+"m.select_answer": {
+    "version": "1",
+    "conf_id": "1674732106391mz4ygIc84Q2Z6mJ5",
+    "call_id": "35657a5b793ce",
+    "party_id": "123456",
+    "selected_party_id": "678910",
+}
+```
+
+Once the client receives this, it decides what tracks it wants to receive and then sends
+a subscribe message over the data channel:
+
+m.call.subscribe
+```
+"m.call.subscribe": {
+    "seq": 1,
+    "media_uuids": {
+        "aaaa-aaaa-aaaaaa-aaaa-aaaa": {
+            "width": 1024,
+            "height": 576,
+        },
+        "bbbb-bbbb-bbbbbb-bbbb-bbbb": {},
+    },
+},
+```
+
+This has also been rearrnaged a little to make the media UUIDs the keys and remove the
+unsubscribe section which is unnecessary if we always send the complete set of tracks we
+want to receive (we unsubscribe by just removing the media UUID from the dict).
+
+This also now contains a sequence number. This is a monotonically increasing integer, starting
+at 0 and scoped to the lifetime of the peer connection. The focus will send a reply containing
+this sequence number to acknowledge that it has processed the message. This can be a positive ack:
+
+m.call.ack
+```
+"m.call.ack": {
+    "seq": 1,
+    "result": "success",
+}
+```
+
+...or an error:
+
+m.call.ack
+```
+"m.call.ack": {
+    "seq": 1,
+    "result": "error",
+    "errcode": "M_UNKNOWN",
+    "error": "Internal server error",
+}
+```
+
+This may give some indication as to why some tracks were not available (should it have errors per
+media UUID, perhaps?)
+
+If the focus needs to renegotiate to send the tracks, it does so, describing the media UUIDs it intends to send on the
+transceivers once the negotiation is complete:
```

Comment:
In this case, thats the transceiver id's (mid's) from the Focus side. Right?

Reply:
They should be the same, no? Since the media lines must be the same on each side.

Reply:
Right, the  media lines are bidirectional and correlate to the remote side. 

What i mean is, whoever initiates the offer determines what is sent over the mid-line. When Focus only receives tracks via a connection, the Focus will never have a reason for start a media renegotiation, because on their side the media will never change. 

In case the Focus start a renegotiation and creates an offer, then the focus sending media as well. In this case the Focus defines, what is sending over the media lines. But this means the client will receive a track from another peer. 

If the client not sending media to the Focus the Focus starts with  0 media line. If the client already sending to the Focus the Focus could use as well the 0 media line. But we decide in this case the Focus will use a new Transceiver. 

Long story short: The media lines alone define not what will sending over it. I know we want use pre defined Transceiver and reserve the 0-media lines and 1-media lines for Client sending medias. However, I have still my doubts that this is a good idea.

Reply:
I'm not sure I understand. From my PoV, the three different scenarios work like this:

If Alice (the client) isn't sending any media to the focus, the focus starts sending media to her and starts a renegotiation to add a transceiver, which becomes mid 0. It includes an `m.call.describe_media` saying that the media on mid 0 is Bob's media.

If Alice is already sending media, the focus renegotiates but decides to add a new transceiver. This becomes mid 2, so its `m.call.describe_media` informs the client that mid 2 contains Bob's media. There is no entry for mid 0 or 1 because the focus isn't sending anything on those transceivers, it's only receiving.

If Alice is already sending media and the focus decides to re-use one of those transceivers, its `m.call.describe_media` informs the client that Bob's media will arrive on mid 0. Alice will re-send her own `m.call.describe_media` in her answer re-iterating that her own media is being sent on mid 0 since she's still sending it in the other direction. The focus's `m.call.describe_media` describes what's flowing on the transceivers in the focus -> client direction and the client's `m.call.describe_media` describes what's going client -> focus.